### ISSUE TITLE
Undocked Mixer width not the same on reopening score.

### ISSUE THREAD

**Issue by scorster** [musescore/MuseScore#21148]
### Issue type

UX/Interaction bug (incorrect behaviour)

### Bug description

[Reports and descriptions located here on .org's forum.](https://musescore.org/en/node/359180)

Not sure I can help further on this ... other than to reflect that [others have reported this MS4.2 Mixer issue as well.](https://musescore.org/en/node/359180)

### Steps to reproduce

**UPDATE (2024-02-09):** 

In the score attached [at the following .org link](https://musescore.org/en/node/359180#comment-1228459) you'll find instructions on how to replicate with the score attached there. (I get varying results, but for some reason the mixer is always resized on reopening, and if my wits are about me, it is always narrowed.)

https://musescore.org/en/node/359180#comment-1228459

Unfortunately **the following steps _DID NOT_ REPLICATE the error:** 

     • Create a new score in MS4
     • undock the mixer and widen it 
     • Save 
     • Quit 
     • Reopen

### MuseScore Version

museScore 4.2

### Regression

Yes, this used to work in MuseScore 3.x and now is broken

### Operating system

MacOS 13.5.2

**Comment by bkunda** [comment]
Closing this for now as we haven't got any clear reproduction steps (and steps that are explicitly marked as "not reproduction steps" are extremely unhelpful). 
Happy to re-open if this issue can be drafted in a clearer, and more helpful way.

**Comment by scorster** [comment]
I still can't replicate Mixer width issue from scratch, but in the score attached at the following .org link, you'll find instructions on how to replicate with the attached score. I get varying results, but for some reason the mixer is always resized on reopening, and if my wits are about me, it is always narrowed.

https://musescore.org/en/node/359180#comment-1228459

**Comment by Daniel-R15** [comment]
The "narrow Mixer width" problem still exists as at Mu 4.4. 
OS: Windows 10 Version 2009 or later, Arch.: x86_64, 
MuseScore Studio version (64-bit): 4.4.0-242390800, revision: github-musescore-musescore-0fcd11b

**Comment by scorster** [comment]
*The "narrow Mixer width" problem still exists as at Mu 4.4.*

The steps to replicate are simple:

Open a score
Undock the mixer
Widen or narrow the mixer width
Toggle the mixer button off, then on
**Result:**  the mixer width reverts to some default minimum. Happens EVERY time.

Note that the "pencil width" minimum-width no longer occurs.

**Comment by dmland** [comment]
"Pencil width" minimum is still happening for me under MuseScore 4.4 (OS: macOS 12.7, Arch.: x86_64, MuseScore Studio version (64-bit): 4.4.2-242570931, revision: github-musescore-musescore-3130f97).

As a workaround, i created a (fragile, hackish) macro for Keyboard Macro that calculates the correct width by probing the app with AppleScript, then positions the mouse along the right-hand edge and drags to reveal all the channels.

**Comment by Jojo-Schmitz** [comment]
I can confirm that issue with the very narrow mixer, happened to me too several times.
I don't know though how to reproduce

**Comment by scorster** [comment]
> I can confirm that issue with the very narrow mixer, happened to me too several times.
> I don't know though how to reproduce

Thanks Jojo. And you're on Windows?

**Comment by Jojo-Schmitz** [comment]
Yes, Windows 11

**Comment by scorster** [comment]
v4.4.4 marks the first time I've not consistently encountered the undocked "pencil width" mixer issue.

However recently I noted the [first time I heard this issue reported by a Linux user.](https://musescore.org/en/node/359180#comment-1271738)

**Comment by scorster** [comment]
## Further observations on undocked Mixer position, after a toggle cycle (close/reopen)

### Moving the Mixer

MuseScore 4.4.4 stores the location of the undocked Mixer when the user closes the document. 

- Open a new MuseScore 4.4.4 score
- Open the Mixer, undocked. 
- Move the Mixer location
- Toggle the Mixer closed and reopened

Then repeat the last three steps placing the undocked Mixer in distinctly different position

**Result:** MuseScore _usually_ remembers the location of the that the Mixer was in when the document was last saved.
Subsequently, moving the mixer and closing it fails to save the last location, but rather the location when the score was last saved, or some prior default.

### Then test after saving and closing:

- place the undocked Mixer in a uniquely memorable location
- Close the score
- Reopen the score
- and return to the first list of steps

### Issues with resizing the undocked Mixer

After widening or narrowing the undocked Mixer, a Mixer toggle cycle usually restores it to a narrower width than when it was last widened. (And as previously stated, the Mixer usually doesn't open with its upper-left corner at the x,y origin of its most recent position.)

**Comment by FrancRos31** [comment]
I found a way to reproduce this issue:
1. open MS
2. with the program empty, open the palettes, instruments and properties panels
3. drag the properties panel on the right of the screen (don't change the three panels sizes)
4. create a score from the grand staff template
5. open the mixer
6. undock the mixer
7. place the mixer like the following:
![1](https://github.com/user-attachments/assets/3709760d-3bc6-42bf-9233-32cd77747e4d)
8. save the score 
9. close the app leaving the mixer open in that position
10. reopen the score from its specific file (it doesn't work if you open the score from the recent files in MS)
Expected behaviour: the mixer is the same width as before.
Actual behaviour: the mixer changed its width.

Now we have the panels as reference points for the width of the mixer.

OS: Windows 10 Version 2009 or later, Arch.: x86_64, MuseScore Studio version (64-bit): 4.4.4-243461245, revision: github-musescore-musescore-2232670

https://github.com/user-attachments/assets/e51746d8-d74a-405d-9872-46c895daf9c5

**Comment by scorster** [comment]
@FrancRos31 

Thanks for replicating this the Mixer-width bug on Windows 10 ... and for illustrating it with a video!

**Comment by FrancRos31** [comment]
@scorster You're welcome! I think this is enough for reopening this issue (@bkunda), but I can't verify for other OS.

**Comment by scorster** [comment]
@Scorster wrote >

    _[Mixer width issue] Extant in MuseScore Studio 4.4.4_

        _• the position of the Undocked Mixer is not remembered_
        _• the width of the Undocked Mixer is not remembered_

Still, in MuseScore 4.5:

     when toggled closed and then toggled open, the undocked Mixer fails to remember its width or position.

Best I can tell, it's been over a year and these matters have not been assigned to a project.

**Comment by krasko78** [comment]
My mixer is hidden as I don't use it but yesterday while I was working on another issue I must have clicked the Mixer button or F10 because I saw the Mixer and it was pencil-wide. It was pretty much reproducible. Now, after I finished what I was focused on, I've started looking into the this issue and I think I have something on it.

Here's what I am seeing so far (we are discussing only undocked mixer):
1. MuseScore seems to remember the position and size only when the mixer is dragged by clicking and dragging its tab (i.e. the text "Mixer"). It does not remember them if you drag the window by clicking on an empty area to the right of the "Mixer" tab.

2. The size and position remembered if you drag by the tab are the ones at the start of the drag. (chuckle)

3. Every time MuseScore starts, the Mixer is shrinked so eventually it gets down to pencil width. This is due to some scaling that is performed which I am currently trying to understand. The first time I undocked the Mixer in my official 4.5.0 installation, I was able to shrink it just by hiding it and showing it multiple times but once I moved it I could not shrink it by hiding and showing it any more, only by restarting MuseScore.

If I find out more, I'll post it.

**Comment by scorster** [comment]
Point 1 is quite interesting. If this pays off I'll give you extra sleuth points!

I tried to confirm on MacOS. But in 4.5 the undocked mixer location is forgotten every time I toggle the Mixer off and back on. The mixer dimension is forgotten too, but not always. 

The occurs with either type of drag:

• when I drag by the Mixer Tab 
• when I drag by the Title bar's various non-tab areas. I never get an exactly reconstructed Mixer.

**Comment by krasko78** [comment]
@scorster

Point 2 is also interesting. Try this:
1. Resize the Mixer as big as you want it to be.
2. Move it to the desired location.
3. Now, drag it again to an arbitrary location by clicking on the "Mixer" tab.
4. From then on, doesn't hiding and showing it keep it at the desired position and size set at steps 1 and 2? (current session only)

Closing and restarting MuseScore should then restore it at the desired location but the size will be smaller.

Another experiment:
1. Follow the above steps.
2. Shrink the main window of MuseScore to the smallest width you can.
3. Close and restart MuseScore a few times. This time the Mixer should grow wider instead of narrower. :)

**Comment by krasko78** [comment]
@scorster & @FrancRos31 Hey guys! I need your cooperation on this one so we can fix it once and for good. I have linked a PR with a fix but the MuseScore team is not convinced. I fully stand behind my comments above until someone manages to prove me wrong If they manage. Or until I manage to convince everyone else. The latter is my preference. :) Let's get down to some business, shall we?

First we should know that there are two separate issues here:
1. When closing and re-opening the Mixer (or any other floating panel for that matter but let's stick to the Mixer) in the same session, it does not remember its position and size.
I have created [PR27695](https://github.com/musescore/MuseScore/pull/27695) that should fix this. This PR makes MuseScore remember the position and size of a floating panel when it is closed by the user.
Without the PR, I stand behind my findings in my erlier comments above: MuseScore remembers the size and position of the Mixer the moment you start dragging it. This effectively means it remembers the penultimate position and size it had before it was moved for the last time. Here's an example: suppose it is located at (50, 50) and its size is (200, 100). Move it once. Close it and reopen it. It will return to (0, 0) with a size of (50, 50). This is the initial position because it was moved only once and this initial position is the penultimate one. Now, drag it to (100, 100) and resize it to (400, 100). Then drag it anywhere else on the screen. Close it and re-open it. It should be restored at (100, 100) with a size of (400, X). X will not likely be 100 because the Mixer auto-resizes itself when shown so we should only focus on its position and width. Can you confirm this for me?

2. Now the second problem which is trickier. The randomly changing width, including the pen-width. I do stand behind my earlier comments here as well. I don't think this has to do with the score - MuseScore saves the layout globally and not per score. My findings are:
- MuseScore resizes all floating panels when re-launched. So the triggering factor is simply closing and re-launching MuseScore. If the width of the MuseScore window is > 1150px, all floating panels will have their widths shrank. And vice versa: if the width of the MuseScore window when closed is < 1150px, when it is relaunched all floating panels will have their widths increased. The further the width of the main window from 1150px. the more pronounced the shrinking/expanding of the floating panels. The same goes for the height except that this reference height is around 768px. This value can vary from platform to platform.

Most users have MuseScore maximized, i.e. larger than 1150x768. This is the reason for the shrinking. I guess most people use laptops with 1920 x 1080 px of resolution. So 1920 : 1150 is almost 2.0, which means every time they close their maximized MuseScore application and re-launch it, the Mixer will be almost 2 times narrower than before. Restart MuseScore a few times and it will quickly go down to pen-width which is the minimum width possible (there is a minimum).

Can you confirm this carefully for me too?

Watching FrancRos31's video on how to reproduce the issue everytime clearly shows that he closes MuseScore and re-launches it which confirms my theory. The team appears to be confused by FrancRos31's comment that the issue does _not_ occur if you reopen the score from the recent files. FrancRos31, when you say it does not happen from the Recent files list, I assume you mean it does not happen if you close the score _without_ closing MuseScore and simply re-open it from the Home tab? If this is so, then it also confirms my theory because in this case there is no re-launch of MuseScore. It would be very strange if something changed when re-opening from the Recent files list because the panels are simply hidden since they belong to the Score tab and are shown back when a file is open and the Score tab is again available.

Any questions let me know. Many thanks!

**Comment by mathesoncalum** [comment]
> in this case there is no re-launch of MuseScore

FYI @krasko78 what I'm talking about [here](https://github.com/musescore/MuseScore/pull/27729#pullrequestreview-2790351618) would still involve a relaunch of MuseScore, but the extra step seems to prevent the bug and I was wondering why.

1. Make sure "Start empty" is selected in Preferences -> General
2. New score, undock the mixer
3. Close MuseScore
4. Relaunch (directly, not by double clicking a file)
5. Open via recent files

With those steps I cannot repro the bug (but I can with the the steps shown in the video).

**Comment by krasko78** [comment]
Thanks, @mathesoncalum There was a lot of information around this issue. That's why I asked the guys for some clarification. I did some further investigation yesterday and I think I have all the answers. I'll post later today in the PR. If you open from Recents after relaunching MuseScore, the widths are OK but the heights are still slightly reduced (at least on Windows). This cannot be observed with the Mixer because it auto-calculates its height but can be observed if you undock the Selection Filter for example. Have a great Friday everyone!

**Comment by FrancRos31** [comment]
Hey, @krasko78 & @mathesoncalum! I will gladly help to solve this very annoying issue. I'll need some time to test everything.  

> 1. Here's an example: suppose it is located at (50, 50) and its size is (200, 100). Move it once. Close it and reopen it. It will return to (0, 0) with a size of (50, 50).

If MS remembers the position of the mixer before you start dragging it, after I reopen, the panel should be positioned at (50, 50), with a size of (200, 100), not p (0, 0) and s (50 ,50). Right, @krasko78? 

Now, to clarify the reproduction steps in my video: select "Start empty" in Preferences -> General, create a file and save it. 

If you close the score window, but not MS, and reopen the file from the Recent files list, the bug DOES NOT occur.

https://github.com/user-attachments/assets/1c309900-c9d5-4a95-80e2-9bc8ee7efe3a

If you leave the score opened, close MS, reopen MS and reopen the file from the Recent files list, the bug DOES NOT occur.

https://github.com/user-attachments/assets/203a79a7-a2af-493b-aca4-41116d7df15e

If you close the score window, close MS, reopen MS and reopen the file from the Recent files list, the bug DOES NOT occur.

https://github.com/user-attachments/assets/b38d1acd-e41f-4c15-abdd-1284d614c0f5

The only way for the bug to occur every time, is to open the file from its specific .mscz after closing MS, no other way to my knowledge.

If you confirm from your side, I'm gonna update the reproduction steps above my video in my previous comment.

**Comment by krasko78** [comment]
Thank you @FrancRos31! The missing piece for me was "Start empty" because I have my MuseScore set to restore the last session and I had completely forgotten about "Start empty". 😃 For me everything adds up perfectly now. Great team work! 👏

**Comment by scorster** [comment]
Extant in MuseScore Studio 4.5.2 on MacOS 13.5.2

With the undocked Mixer:

   • Move it's location
   • Toggle it closed
   • Toggle it back open

   Result: The Mixer appears in a different location

However if I *resize* the undocked mixer and close/reopen sometimes it reopens at the previous size and location.

**Comment by krasko78** [comment]
@scorster That's correct. This part of the issue (there are two subissues here) will be fixed in [PR27695](https://github.com/musescore/MuseScore/pull/27695) which is still waiting for a review.

**Comment by scorster** [comment]
Ah, thanks! 

I thought the previous "closed as completed" referred to the central issue of this post. 

Thanks for your work on these persistent bugs!
